File: src/198_House_Robber/main.py
# coding=utf-8

class Solution(object):
    #Runtime: 20 ms, faster than 69.13% 
    #Memory Usage: 13.4 MB, less than 6.26%
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        size = len(nums)
        if size == 0:
            return 0
        elif size == 1:
            return nums[0]
        p2, p1 = 0, nums[0]

        # 状态方程 dp[i] = max(dp[i-1], dp[i-2]+nums[i])
        for i in range(1, size):
            p2, p1 = p1, p2
            p1 = max(p2, (nums[i] + p1))

        return p1

if __name__ == "__main__":
  gua = Solution()
  nums = [1,2,3,1]
  rtn = gua.rob(nums)
  print("expect value = 4, actual value = %d" % (rtn))

File: src/198_House_Robber/main.cpp

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    //Runtime: 0 ms, faster than 100.00%
    //Memory Usage: 7.7 MB, less than 68.31%
    int rob(vector<int>& nums) {
        // 和股票问题一样的两状态问题，一个状态是是否抢nums[i]，另一个状态是i
        // 可以迭代+空间压缩，使得空间复杂度为O(1)
        int rob = nums[0], nrob = 0;
        for (int i = 1; i < nums.size(); ++i) {
            int tmp = rob;
            rob = nums[i] + nrob;// rob定义的是最后一家一定要抢的最大值
            nrob = max(tmp, nrob);// nrob定义的是最后一家不抢的最大值
        }
        return max(rob, nrob);
    }
};

int main() {
  Solution gua;
  int rtn;
  vector<int> nums;
  rtn = gua.rob(nums={1,2,3,1});
  cout << "expect value = 4, actual value = " << rtn << endl;

  rtn = gua.rob(nums={2,7,9,3,1});
  cout << "expect value = 12, actual value = " << rtn << endl;

  return 0;
}


